### ISSUE TITLE
Creating a QEMU VM will fail with OOM when it could use memory used by ARC

### ISSUE THREAD

**Issue by nagisa** [openzfs/zfs#6427]
### System information
Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       | nixos
Distribution Version    | unstable
Linux Kernel                 | 4.12
Architecture                 | x86_64
ZFS Version                  | 0.7.0-rc4
SPL Version                  | 0.7.0-rc4

### Describe the problem you're observing

Consider a system with 16GB of RAM. ARC on such system is some 7.8G in size. Starting a QEMU VM with 10GB of RAM will fail with an error indicating allocation failure:

```
$ qemu-system-x86_64 -m 10000
qemu-system-x86_64: cannot set up guest memory 'pc.ram': Cannot allocate memory
```

To make this work, it is necessary to make ARC smaller, for which it is necessary to use incremental pressure (AFAICT anyway):

```
$ python -c 'i = [bytearray(11*1024*1024) for i in range(1024)]'
$ qemu-system-x86_64 -m 10000
// VM runs
```

This does not happen on the traditional filesystems where the cache is accounted for in buff/cache. Furthermore, if the VM doesn’t actually touch the memory, it will not even evict the caches.

### Describe how to reproduce the problem

* Fill ARC.
* Run `qemu-system-x86_64 -m (freemem+some_arc)`

**Comment by gmelikov** [comment]
I have this behavior in many other scenarios too, when app needs to check free space before allocation (VM start, games textures load, etc).

**Comment by zviratko** [comment]
I hit this recently as well, I had to decrease zfs_arc_max, spawn the VM and then increase it back again.
This is because ZFS doesn't use the same cache as other filesystems do, but has it's own (in SPL?) which allocates memory on the side.
Would be nice to be able to reclaim ARC memory in the same way, though...

**Comment by behlendorf** [comment]
Also potentially related to #6506.

**Comment by dbavatar** [comment]
@behlendorf Yes looks like the same issue as #6506.

**Comment by dweeezil** [comment]
I've certainly seen this in the context of Qemu/KVM VM guests as well and will concur with @gmelikov's observation that this can occur in other scenarios.  One of my "junk programs" [maxsbrk.c](https://github.com/dweeezil/junkutils/blob/master/maxsbrk.c), which I use on occasion to sweep away memory used by ZFS or other applications, can also be used to demonstrate the issue.  As best as I can tell, the issue has to do with the manner in which Linux handles allocations when ```overcommit_memory=0```.  If I try to allocate X bytes of memory in one fell swoop, it may fail, but if I simply allocate it in, say, 10 chunks of X/10 bytes, it will likely work.  Qemu/KVM likely attempts to allocate its guest memory in one chunk so it can easily fail.  Setting ```overcommit_memory=2``` is typically not an option, at least for desktop environments.  I've yet to find a good solution to this issue but haven't really dug into the heuristic used in ```overcommit_memory=0``` mode.

**Comment by gmelikov** [comment]
@behlendorf it may be related to #6506, but I always had this issue on 0.6.5.8-0.6.5.11 too.

**Comment by nagisa** [comment]
I recently encountered this with a recent versions of kernel (4.15) and zfs (0.7.8) when trying to compile a large project during the linking step.

After my usual `[bytearray(6*1024*1024) for i in range(1024)]` in python the linking succeeded.

**Comment by devZer0** [comment]
i'm getting this on proxmox ve 5.2 . that is an debian + zfs based "basic appliance/hypervisor-os" to run virtual machines. 

the machine has 12gb of ram and after running some file-system testing, scrub etc in the main os, i cannot even start a single qemu/kvm VM with 4gb ram as ram usage is at 71% (no vm is running) 

i think running/starting VMs is more important then having ARC being kept in ram.  

IMHO, arc should be dropped or at least reduced by some gigabytes when some process getting started.

one could say proxmox should adress this in their distro somehow, but i don't really get the point why a filled arc would prevent any process to get started/executed (as buffers/caches also get dropped when processes/tasks getting executed)

**Comment by beren12** [comment]
Did you even read the comments? Arc has no problem releasing the memory, but the vm asks for it all at once, and zfs needs a sec. One thing you can do is lower the arc size by as many gb as you need, then after the vm starts set it back to 0 or whatever it was set to before. If you really don't care about arc you can drop caches and that will free *all* cached ram on your machine, then start your vm and the caches will slowly refill.

**Comment by devZer0** [comment]
ok, after re-reading i think i get the point now. thanks. 

so, do you think the correct way to proceed is to make qemu (and other memory hogs which may fail) "zfs arc aware" so they will ask for free memory in a "more friendly manner" so zfs arc has more time to release free mem ?

**Comment by devZer0** [comment]
https://github.com/zfsonlinux/zfs/issues/6740

**Comment by devZer0** [comment]
i found that in proxmox the problem happens because the system is configured to have no swap by default - because of deadlock when haveing swap on zvol.   the oom went away when adding swap, even with swappness=0

as the system is software-raid-only i will need to add mdadm raid1 volumes on the zfs disks as a workaround

https://git.proxmox.com/?p=pve-installer.git;a=commit;h=7504ab82e4e2c7698ded503025cfbddaf507dcb4

**Comment by nagisa** [comment]
Having added a little swap I still encounter issues creating VMs, so it is not a solution for me.

**Comment by beren12** [comment]
Well yes, you would need enough swap that the free ram + swap was at least as much as the vm needed…

**Comment by kpande** [comment]
you should be using hugepages.

**Comment by devZer0** [comment]
is anybody here using kvm via recent proxmox 6.1 and can still reproduce the issue there?

**Comment by devZer0** [comment]
i tried to reproduce this issue on recent proxmox 6.3 system which is using ZoL 0.8.5 , but i did not success, so apparently something must have changed the previous behaviour.  maybe it's fixed, maybe it's harder to reproduce on a system with more ram ressources (as the system where it happens was more memory constrained)

can anybody give a comment if this still happens with recent kvm/ZoL versions ?

**Comment by bghira** [comment]
yes it still happens in OpenZFS 2.0

**Comment by devZer0** [comment]
on my up-to-date proxmox  server with 70GB of ram, after filling ARC to occupy almost whole of the available ram ,
i can resume a VM from disk with 16GB (filled) RAM and  and do even more pressure to the arc with "fillmem.py 16000" in parallel.  no out-of-memory errors and no OOM killer, it just works. i see the arc collapse as expected.

zfs-2.1.1-pve3
zfs-kmod-2.1.1-pve1
pve-manager/7.1-7/df5740ad (running kernel: 5.13.19-2-pve)

@bghira , any news from you?  what is your platform ?

`
#!/usr/bin/env python

import sys
import time

if len(sys.argv) != 2:
    print "usage: fillmem <number-of-megabytes>"
    sys.exit()

count = int(sys.argv[1])

megabyte = (0,) * (1024 * 1024 / 8)

data = megabyte * count

while True:
    time.sleep(1)
`

**Comment by amotin** [comment]
@moritztim Because of lack of activity.  And because nobody care about version 0.7.0.  And because there were plenty of ARC-pressure changes in 2.2 and 2.3.

**Comment by amotin** [comment]
@moritztim I'm sorry to hear, but what do you expect to happen without you providing any additional details, like ZFS version, memory and ARC stats, errors, etc?  As I have told, any reports on this topic before at least later 2.2 are likely irrelevant.

**Comment by moritztim** [comment]
I see, I assumed it must be the same exact issue, but if it's supposed to be fixed, I'll open a new issue with more details. I don't currently have the time for that but hopefully I'll get around to it
